File: Java/src/_0149_Max_Points_on_a_Line/Solution.java
package _0149_Max_Points_on_a_Line;

import javafx.util.Pair;

import java.util.HashMap;

public class Solution {
    public int maxPoints(int[][] points) {

        int res = 0;
        for (int[] p : points) {
            int ss = 0, vs = 0;
            HashMap<Double, Integer> cnt = new HashMap<>();
            for (int[] q : points) {
                if (p == q) ss++;
                else if (p[0] == q[0]) vs++;
                else {
                    double k = (double) (q[1] - p[1]) / (q[0] - p[0]);
                    cnt.put(k, cnt.getOrDefault(k, 0) + 1);
                }
            }
            int c = vs;
            for (double k : cnt.keySet()) c = Math.max(c, cnt.get(k));
            res = Math.max(res, c + ss);
        }
        return res;
    }

    public static void main(String[] args) {
        int[][] point = {{1, 1}, {3, 2}, {5, 3}, {4, 1}, {2, 3}, {1, 4}};
        System.out.println((new Solution()).maxPoints(point));
    }
}

File: Python/_0149_Max_Points_on_a_Line/Solution.py
# coding=utf-8
# Date: 2021/6/24 10:56
from typing import List


# 执行用时：88 ms, 在所有 Python3 提交中击败了26.28%的用户
# 内存消耗：14.8 MB, 在所有 Python3 提交中击败了91.91%的用户
class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        res = 0
        for p in points:
            ss = 0
            vs = 0
            cnt = {}
            for q in points:
                if p == q:
                    ss += 1
                elif p[0] == q[0]:
                    vs += 1
                else:
                    k = (q[1] - p[1]) / (q[0] - p[0])
                    if k in cnt:
                        cnt[k] += 1
                    else:
                        cnt[k] = 1
            c = vs
            for k, v in cnt.items():
                c = max(c, v)
            res = max(res, c + ss)
        return res


if __name__ == "__main__":
    print(Solution().maxPoints([[1, 1], [2, 2], [3, 3]]))  # 3
